accounts-db/src/accounts_db.rs

```diff
@@ -6809,27 +6809,22 @@ impl AccountsDb {
         &self,
         slot: Slot,
         accounts_and_meta_to_store: &impl StorableAccounts<'b>,
-        txn_iter: Box<dyn std::iter::Iterator<Item = &Option<&SanitizedTransaction>> + 'a>,
+        mut txn_iter: impl Iterator<Item = &'a SanitizedTransaction>,
     ) -> Vec<AccountInfo> {
-        let mut write_version_producer: Box<dyn Iterator<Item = u64>> =
-            if self.accounts_update_notifier.is_some() {
-                let mut current_version = self
-                    .write_version
-                    .fetch_add(accounts_and_meta_to_store.len() as u64, Ordering::AcqRel);
-                Box::new(std::iter::from_fn(move || {
-                    let ret = current_version;
-                    current_version += 1;
-                    Some(ret)
-                }))
-            } else {
-                Box::new(std::iter::empty())
-            };
+        let mut write_version_producer = if self.accounts_update_notifier.is_some() {
+            let current_version = self
+                .write_version
+                .fetch_add(accounts_and_meta_to_store.len() as u64, Ordering::AcqRel);
+            current_version..current_version + accounts_and_meta_to_store.len() as u64
+        } else {
+            0..0 // dummy empty iterator
+        };
 
-        let (account_infos, cached_accounts) = txn_iter
-            .enumerate()
-            .map(|(i, txn)| {
+        let (account_infos, cached_accounts) = (0..accounts_and_meta_to_store.len())
```

Comment:
if the txn_iterator is empty, we still want to iterate over the accounts to store.

Might make sense to just have some sort of iteration over `accounts_and_meta_to_store`?

Reply:
Hm sounded good in theory, but when I just took a stab it I ran into a problem immediately.
We implement `StorableAccounts` on various tuple types. Rust does not allow us to implement `Iterator` for arbitrary types that we didn't define.

So afaict we'd need to wrap these in some type or have an iter fn - which doesn't seem much better imo.